\chapter{Squiggle Discretization and Similarity Between Squiggles}

\label{kap:proposedMethod}

As we stated in the previous chapter, we will work directly with the squiggles instead
of the basecalled sequences. However, for this to be possible we need to have 
some reference squiggle that we can sample and use as an index in which
we will search for the squiggles once we are sequencing.
Unfortunately, we are given the reference only in the form of a DNA sequence.
In this chapter we will show how to create the reference squiggle from the reference DNA sequence and also
introduce more suitable representation of the signal that enables easy searching of the
query signal in the reference.

\section{Squiggles as Sequences of Discrete Observations}
\label{section:squigglesAsDiscrete}

Given a simulated squiggle from the reference DNA, we want to decide whether the
squiggle passing through the pore has originated from the reference. This task
can be reformulated as finding if the current squiggle is similar to a segment of the reference squiggle.

We propose to discretize the squiggles to facilitate their better representation
and also provide easy searching capabilities in this representation. Note that, we are
not very limited by the preprocessing time, but very fast processing is required during
the sequencing of the DNA. This puts certain speed limitations on our discretizing
process during the actual sequencing of the DNA.

We will represent the squiggle as a string of characters. We split the squiggle
vertically into several windows so that everything between the minimum and maximum
value is allocated to one of these windows. The windows are of a constant width and do not
overlap. More formally, let $a_i$ be the readout at the time $i$ and $m$ be the number of
vertical windows. Denote the squiggle $s$ of length $n$ as $s=a_1a_2a_3\cdots a_n$.
Besides this, we are given values $min_a$, $max_a$ that $\forall a_i: min_a \leq a_i < max_a$.
We say that the readout $a_i$ is in $j$-th window if it holds that:

\begin{center}
$min_a + j\cdot \frac{max_a-min_a}{m} \leq a_i < min_a + (j+1)\cdot \frac{max_a-min_a}{m}$
\end{center}

Let the readouts $a_0, a_1, \cdots, a_n$ correspond to the windows $w_0, w_1, \cdots ,w_n$
respectively. $ls$ such that $ls=w_1w_2\cdots w_n$ is called \textit{level sequence}.
Level sequence in which the subsequent appearances of the same character are substituted
by the single apperance of that character is called the \textit{level string} of the squiggle $s$.
We will call the transformation from the level sequence to the level string a \textit{contraction}.
So if the $ls=aabccaa$ then the level string from this level sequence is "abca". An example of the transformation
of the squiggle to the level string for $w=6$ is shown in Figure \ref{obr:levelsCreation}.

\begin{figure}
\centerline{\includegraphics[width=0.8\textwidth, height=0.4\textheight]{images/levelsCreation}}
\caption[TODO]{Squiggle with the level string displayed on the x-axis}
\label{obr:levelsCreation}
\end{figure}

Discretization has several important properties that will be later used.
First, it is deterministic and the same squiggle has the same level
string each time for the unchanged $w$. Second, small changes and variance in the signal readouts
are unlikely to change the resulting level string most of the time. Note that the squiggle
consists of events that are represented by a longer signal that is around the same level.
With this method, we hope that each event will remain insde a single window over its
duration. Subsequent contraction of the level string thus results in one event being represented by
a single character of a level string in most cases.

Changes to the parameter $w$ balance specificity and informativness of our method.
With the $w=2$, we cannot say a lot about two squiggles that have the same level string.
With larger $w$, we can miss two squiggles from the same DNA sequence if there is
a too much noise present in one of them.

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.3\textheight]{images/levelDiff}}
\caption[TODO]{Examples of the squiggle divided into 3 and 12 windows}
\label{obr:levelDiff}
\end{figure}

Consider two scenarios, when the number of levels is small and high.
Figure \ref{obr:levelDiff} shows that for the small number of levels, a lot of
different events stay in the same center window even if the higher number of levels
could distinguish between these individual events. With a high number of levels,
like we see in Figure \ref{obr:levelDiff} one event is less likely to remain within the same window.
We will address the signal oscillations in more detail in Section \ref{subsection:oscillations}.

\section{Simulating Squiggles from the Reference}

We already stated that before the process of the selective sequencing we are most
of the time only given the reference DNA sequence. If we want to work directly with the squiggles,
we need to obtain the reference squiggle. We need this reference squiggle
so that we have some reference which we can sample and use later to split squiggles
between that we are interested in and not interested in.

We create a simulated squiggle based on the 6-mer model,
where each subsequent 6-mer in our reference DNA will generate a signal (event) equal to
the mean of the gaussian distribution for that particular 6-mer. This mean value
is the information provided by the manufacturer of MinION through the $k$-mer model
that we mentioned in Section \ref{section:dnaSequencing}.

Our simplistic approach to signal simulation does not take into account that the
signal from one nucleotide is measured several times. Thus, we replicate every entry in the simulated signal
10 times.

We now compare the simulated squiggle to the real squiggle that arose from the real
sequencing. Moreover, we do not want to compare the simulated squiggle with unrelated squiggle
but with the squiggle that corresponds to the same DNA region.

We will look now how to obtain the pair of the simulated-real signal that originate
from the same DNA region. First, we take some real squiggle.
Then, we find the part of the DNA reference that this squiggle corresponds to. After
successfully finding this part of the DNA, we take it and simulate the squiggle from it. We will use
two ways how of finding the DNA region corresponding to the squiggle in this work. The first, more accurate method is to
use Nadavca. This tool, introduced in Section \ref{section:currState} takes
as the input the reference DNA sequence and the particular squiggle we want to find
in it. It then returns the table that accurately maps the nucleotides to the individual
events in the squiggle. This possibility is very accurate and gives a lot of information
that we do not need most of the time such as the mapping between the nucleotides
and events. We are interested only in location of the start and end of the squiggle
in the reference. The second option is to take the read instead of squiggle and use information from it.
Hovewer, we know that the DNA sequence in read is generated by the base-caller algorithm so it has some error
in it. To obtain the sequence of the nucleotides corresponding to this read without
errors, we need to use some algorithm that is able to find this erroneous DNA
sequence in the reference DNA sequence. For this purpose, we use minimap2 algorithm
that we describe more closely in the Section \ref{section:alignMinimap}. This is faster
process as the solution using Nadavca and we will use this option anytime when we need
speed and do not need any additional information that Nadavca provides.

Figure \ref{obr:simVsReal} compares the simulated squiggle to the squiggle obtained
by sequencing the same region of the DNA.

\begin{figure}
\centerline{\includegraphics[width=0.7\textwidth, height=0.3\textheight]{images/simulateRef}}
\caption[TODO]{Comparison of the real (upper graph) and simulated (bottom graph) signal with the corresponding nucleotid sequence on $x$-axis}
\label{obr:simVsReal}
\end{figure}

Now we have a simulated reference squiggle. We can see on Figure \ref{obr:simVsReal} that the simulated and real
squiggle are not identical. This poses a problem as we want to have these squiggles as
similar as possible. The first thing that we want to consider is normalization.
One of the most wide-spread ways of doing this is to subtract mean and dividing
the signal by the standard deviation. The other possible solution is subtracting median value,
as the frequent outliers can deform the mean. We will stick to the first solution,
as our experiments showed that there is no big difference between squiggle mean
and median value in most of the squiggles.
After the normalization, we remove outliers in a way that everything over and under certain values will
be clamped to the target range. Most of the time, this will be -2 as the lower bound
and 2 as the upper bound.

Another visible problem can be seen when we compare events in the simulated and real squiggle.
The events in the simulated signal are perfect repeats of the same value. In the real signal, we can sometimes see
the phenomenon called \textit{drift}. This is a fact that real event tends to slide
up or down over time. Another difference is that real squiggle
is sometimes contracted in some places. This is caused by the fact that the DNA
molecule is not moving through the pore at a constant speed.

There are two possible ways how to overcome the differences between real squiggle
and simulated squiggle. One approach is to make the simulated squiggle more like
the real squiggle, including adding a noise and a varying event deviation. In fact there
are more advanced squiggle simulators such as DeepSimulator\cite{deepsimulator}.

The opposite approach is to make the real squiggle less noisy. We decided to go the
second way so we will now address the most important differences between the
real and simulated squiggles.

\section{Signal Oscillations and Continuity}
\label{subsection:oscillations}

Figure \ref{obr:simVsReal} clearly shows that one of the biggest differences between simulated and
real squiggles is the noise which can be seen as small oscillations of the signal. This is
one of the reasons that our discretization method use windows - so small oscillations within the window
are dealt with. We can also easily adapt the number of windows.
However, when our method is used without any modifications it
produces much longer level strings for the real squiggles than for simulated squiggles.
The problem arises when the oscillations are located on the borders of the windows
and produce too long level string of the form "cdcdcdcd".
To address this problem, we tried several smoothing techniques to reduce this oscillations. \textit{Moving average} is the
smoothing technique that for every readout $s_i$ in some signal $s$ calculate its new value
$snew_i$ as $snew_i = avg(s_{i}, s_{i-1}, s_{i-2}, \dots , s_{i-k+1})$. We will call
$k$ the \textit{smoothing parameter}. We can also simply change
the average function to the median to obtain the \textit{moving median}. For now, we will stick
to faster yet simpler smoothing techniques.

Another difference between the real signal level string
and the simulated signal level string is that the real signal is more continuous instead
of a simulated signal that jumps. Again, we can use the moving average. Instead of
the moving median, it has a nice advantage that it causes signal not to jump from
one level to another but rather move continuously.

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.5\textheight]{images/testSmooth}}
\caption[Hehe]{Comparison of simulated and real squiggle smoothed by different techniques}
\label{obr:testSmooth}
\end{figure}

In Figure \ref{obr:testSmooth} we can see part of the simulated and corresponding real squiggle.
Then, in the subsequent two graphs, we apply the moving average with a smoothing parameter
equal 5. We can see that this smoothed the simulated squiggle but also smoothed the noise
that we could observe in the events. With the median smoothing, the outcome is quite
similar but lacks the quality of smoothing the simulated signal.

In Figure \ref{obr:anti_zigzag} we see the result of our smoothing techniques
applied on the simulated reference squiggle and real squiggle of lengths $10\,000$. We can see how
big impacts the smoothing techniques have on the lengths of the resulting level strings.
In most of the cases, the majority of level strings from the real squiggles without smoothing
are more than half time bigger in size. After the smoothing we can see big changes in
the distribution of the length ratios.

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.35\textheight]{images/anti_zigzag}}
\caption[Hehe]{The impacts of the smoothing on the underlying squiggles of length $10\,000$. We see the individual 
frequency of reads level string length to reference level string length}
\label{obr:anti_zigzag}
\end{figure}

\section{Identification of Reads Based on Shared $k$-mers}
\label{section:readIdentification}

In this section, we will demonstrate how to distinguish between the cases when two squiggles
originate from the same sequence and cases when we are comparing unrelated squiggles.
We take the pair real squiggle - simulated reference squiggle from the same DNA region and one unrelated real squiggle.
We will then try to distinguish which of the two real squiggles is from the same region as the simulated squiggle
and which is not based only on the level strings derivated from the respective squiggles. We
emphasized the speed many times, so we will try to come up with a fast test.
We split the individual level strings into overlapping $k$-mers (any $k$ subsequent
characters from the string). The size of the overlap between $k$-mers from the reference
squiggle and the sample squiggles should distinguish which squiggle is from the reference and which is not.

\section{Experimental Evaluation}

In our experiments, we want to test:

\begin{itemize}
    \item How accurately the discretization method that we proposed in Section \ref{section:squigglesAsDiscrete} represents the squiggle.
    \item How the proposed identification from Section \ref{section:readIdentification} works on real data
\end{itemize}

In both of these cases we are interested in the results for the different set of parameters
such as the number of levels, length of the $k$-mers and impact of the methods that we
discussed in Section \ref{subsection:oscillations}.

\subsection{Experimental Data}
\label{section:data}

Data that we use for these experiments come from two organisms.
One is \emph{Saprochaete ingens} (dataset sapIng) and
the other one is \emph{Saprochaete fungicola} (dataset sapFun). These
organisms were sequenced using R9.4.1 MinION flow cells at the Department of Biochemistry
of the Faculty of Natural Sciences. We have also obtained the reference sequences of
both of these organisms that were previously published \cite{hodorova2019genome} \cite{brejova2019genome}.
We base-called the raw squiggles using the Albacore basecaller provided by Oxford Nanopore Technologies,
company that also develops the MinION sequencer.

\begin{table}
% v tabulke sa popis zvykne davat nad tabulku
\caption[TODO]{Basic characteristic of the used datasets}
%id tabulky
\label{tab:datasetChar}
\begin{center}
\begin{tabular}{lccccc}
\hline 
Name & \# reads & \specialcell{mean read\\length} & \specialcell{median read\\length} & \specialcell{mean squiggle\\length} &  \specialcell{median squiggle\\length}\\
\hline
sapIng & $3\,000$ & $14\,663$ & $6\,965$ & $151\,002$ & $71\,964$ \\
sapFun & $3\,000$ & $10\,912$ & $4\,775$ & $113\,550$ & $51\,266$ \\
\hline
\end{tabular}
\end{center}
\end{table}

To see if these two organism do not share some big similarities, we align the reads from
\emph{Saprochaete fungicola} to the reference sequence of \emph{Saprochaete ingens} using the Minimap2
algorithm. We run this test on the 500 randomly selected reads from each of these organism. We studied
the ratio of the reads that were successfully matched in the reference sequence. Figure \ref{obr:align_stat}
shows the number of reads that were alligned with the respective percentage of their length that was successfully aligned.
Observe that only 12 reads out of 500 in case of the \emph{Saprochaete fungicola} had more than 50\% of their length aligned.

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.4\textheight]{images/align_stat}}
\caption[TODO]{Number of reads that aligned at least on $x$\% of their length to the reference DNA of the \emph{Saprochaete ingens}}
\label{obr:align_stat}
\end{figure}

For our experiment, we use only the squiggles from the \emph{Saprochaete ingens} if they
are successfully aligned into the reference using the Nadavca. We then create the
simulated reference signal from the reference part that Nadavca identified and pick
some random read from the sequencing of the \emph{Saprochaete fungicola} and run our analysis.
We run this cycle many times over. As the ultimite goal of our work is to achieve
the ability of selective sequencing, we must also restrict ourselves to only work with
the beginning part of the squiggle. We reflect this in our experiments so we
work only with the first $5\,000$ readouts from the squiggle segment that Nadavca identified
as matching to our reference.

\subsection{ROC curve}

A receiver operating characteristic curve is a graphical illustration of how well
the \textit{binary classificator} performs on the data for a particular threshold. The binary
classifier system is a system that tries to predict the binary output from the inputs.
We can look at our experiment as predictor that predicts based on the number of hits, if the squiggle is from the
simulated reference squiggle or not. We can choose some threshold value $t$ and say that
our classificator will predict all the squiggles with the number of hits bigger or equal
than $t$ as coming from the reference squiggle. When working with binary predictor there
are two important rates \textit{true positive rate} TPR and \textit{false positive rate} FPR. 

TPR is a ratio of the correctly classified positive squiggles to the number of the all squiggles.

FPR is a ratio of the squiggles that we incorrectly classified as a belonging to the reference squiggle.

The ROC is curve that represents the dependency of TPR on FPR for various threshold values $t$.
If the binary classificator was random generator the ROC curve would be roughly
linear. Very good classificator obtains the curve that is very close to the point
$FPR = 0, TPR = 1$. We can see the example of a curve that represents very good
binary classificator on Figure \ref{obr:roc}. The binary classificator that randomly
guesses his output is expected to have an ROC curve somewhere around the green line.

\begin{figure}
\centerline{\includegraphics[width=0.5\textwidth, height=0.3\textheight]{images/roc}}
\caption[TODO]{In this picture we see the example of a ROC curve. Green curve represents very good
binary estimator.}
\label{obr:roc}
\end{figure}

\subsection{Level string alignment}
\label{section:alignment}

To see how two level strings compare to each other we will align them. The input
to the alignment algorithm are two strings $s_1$, $s_2$:
$s_1=a_1a_2\cdots a_n$, $s_2=b_1b_2\cdots b_m$. The alignment are strings
$s_3$, $s_4: s_3 = c_1c_2c_3\cdots c_k$, $s_4 = d_1d_2d_3\cdots d_k$ such that:

\begin{enumerate}
\item $s_3$ and $s_4$ was created from $s_1$, $s_2$ respectively by inserting dashes
\item $c_i \neq d_i$ then $c_i = - \lor d_i = -$
\end{enumerate}

Example of allignment of $s_1 = AAACTGC$ and $s_2 = AACGTC$ is:

\begin{center}
$s_3 = $AAAC\,-\,TGC\\
$s_4 = $AA\,-\,CGT\,-\,C
\end{center}

There are a lot of alignments between the two strings. Most of the time we
are interested in some particular alignment. For this purpose, we will create a
scoring system and we will try to find the maximal (minimal) alignment in this scoring
system. We can, for example, minimize the number of dashes. Using the alignment
defined this way we can see how two-level strings compare to each other and how similar they are.
Finding the best alignment is not trivial. We use dynamic programming for
solving this problem. Let $T[i][j]$ be the best alignment of string $s_{1_i} = a_1a_2\cdots a_i$
and $s_{2_j} = b_1b_2\cdots b_j$. We will for simplification say that $\forall i, j: T[i][0] = i, T[0][j] = j$.

\[
T[i][j] = \bigl.
  \begin{cases}
    0, & \forall i,j : i\leq 0 \lor j\leq 0 \\
    max(T[i-1][j-1] + 1, T[i-1][j], T[i][j-1]) & \text{only if} \, a_i = b_i \\ 
    max(T[i-1][j], T[i][j-1]) & \text{otherwise}
  \end{cases}
\]

\subsection{Results}
\label{section:results}

TODO: Pred odovzdanim nahodit nove pekne grafy.

We performed three experiments that relate to the goals that we set at the start
of this Chapter. We use the squiggles from the datasets introduced in Section
\ref{section:data}. We stated that the selective sequencing must be done using
the information from the start of the squiggle. In this all three experiments, we always use
the $5\,000$ readouts from the any squiggle mentioned. 

The first experiment evaluate the number of shared $k$-mers between the simulated
reference squiggle and real squiggle corresponding to it. To put the number of shared
$k$-mers into perspective, we will also look at the number of shared $k$-mers between the
same simulated reference and the unrelated squiggle. For easier description, we will
label the simulated reference squiggle the reference, real squiggle that corresponds
to this reference the positive squiggle and the random unrelated squiggle the negative squiggle.
In all individual testcases we calculate the ratio of the hits of negative squiggle
to the hits of the positive squiggle. We look at the distribution of these ratios.
The ideal case is that most of the testcases will have the ratio as small as possible. We will
track ratio only up to ratio 2:1. If the number of hits of the positive squiggle is $0$,
then this ratio is undefined so we will mark this as a ratio 2:1. The goal of this
experiment is to see how informative is the number of shared $k$-mers regarding the underlying squiggles.

The results of the experiment with using all the smoothing methods that we introduced are on Figure \ref{obr:res_3}.
In this figure, we see the cummulative ratio of the measured ratios.
We see that there are parameters that obtain the ratio $1/2$ for more than 80\% of testcases. 
There is phenomenon of the lines that start with the perfect record and then quickly
go to the right part of the graph. This happenes to the curves representing
the high $k$-mer number. That is because the number of hits in random squiggle is ussually zero
for high $k$ but it is also hard for the real squiggles to obtain
high number of hits so for a lot of positive squiggles, the number of hits is 0 which
shifts the curve to the left. In the Figure \ref{obr:res_3_nonorm} we see the results
for the squiggles not processed by our smoothing techniques. We can clearly see that
the results are way worse.

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.4\textheight]{images/res_3}}
\caption[TODO]{Cummulative distribution of hit ratios}
\label{obr:res_3}
\end{figure}

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.4\textheight]{images/res_3_nonorm}}
\caption[TODO]{Cummulative distribution of hit ratios without smoothing}
\label{obr:res_3_nonorm}
\end{figure}

The second experiment is motivated by the hidden drawback of the first experiment. Consider
two testcases in the first experiment. If the individual two positive squiggles have
the bigger number of hits than their two negative counterparts in respective testcases,
we consider it as a success in the first experiment. However, there can be problem
that even if the both of these testcases are considered successful
the number of hits of the negative squiggle in the second testcase can be bigger than
number of hits of the positive squiggle in the first testcase. This will cause that
even if we have two successful testcases in the first scenario, we cannot predict
if the squiggle is from the reference based on some threshold value of hits. This
can happen for example if a lot of squiggles have considerably lower number of hits
because of the way how we discretize the signal. 
The second experiment look at how often this happens and if the number of hits in
positive squiggles is globally bigger than number of hits in the negative squiggles.
We will again take the reference squiggle, positive squiggle that comes from the
reference segment and a randomly chosen negative squiggle from the other sequencing.
We plot the data using the ROC curve. Note that if two squiggles have the same number
of the hits, we favour the negative squiggle to present objective results.

In Figure \ref{obr:res_2} we see the ROC curve that represents how overall number
of hits in the positive squiggles compares to the number of hits of the negative
squiggles. We can see very good results for the lower number of levels and higher
length of $k$-mers. In Figure \ref{obr:res_2_nonorm} we can see that without smoothing
techniques applied the lower lengths of $k$-mers have better results. This is
the case because it is very unlikely for the unsmoothed squiggle to obtain the longer
match due to the reasons that we worked on in the Section \ref{subsection:oscillations}.

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.4\textheight]{images/res_2}}
\caption[TODO]{ROC curve of the hits with smoothing}
\label{obr:res_2}
\end{figure}

\begin{figure}
\centerline{\includegraphics[width=1.0\textwidth, height=0.4\textheight]{images/res_2_nonorm}}
\caption[TODO]{ROC curve of the hits without smoothing}
\label{obr:res_2_nonorm}
\end{figure}

The third experiment will focus on mutual alignment of the level strings coming from
the real squiggle and reference squiggle respectively. Again, as in previous experiments
we will add random squiggle to see the differences between alignments.
We will align them using the scoring scheme we already presented in Section \ref{section:alignment},
minimizing the number of the dashes. The reason why we do this experiment is that
the shared $k$-mer count only shows the (total) number of shared $k$-mers but it
does not take into account the relative positions of matched pairs. This means
that we can easily count as the matching pair also two $k$-mers that
do not correspond to the same signal but happened to be in the same level string.
Hovewer, for a reasonable number of levels, this will not happen often for a short read.
The big advantage of the alignment of the strings is that it tells us more accurately
what is the overall similarity of the level strings. What we will measure is
how often we can see the successive number of dashes called \textit{gaps}. From
various experiments, we learned we need to distinguish between the length of the particular gaps. We will call the number
of succesive dashes a \textit{gap length}. We split the gaps into the two groups according
to their length. We will create group with the gaps shorter than 6 and the other with
gaps longer than 9. We will count the individual number of gaps in these two groups.

In the Figure \ref{obr:res_1} we present our findings for positive
and negative squiggles. We present the ratio of the number of the gaps of given length divided
by the length of the level string of the respective squiggles. For better visualization,
we show the gaps per 100 level string characters for gaps shorter than 6 and gaps
per $1\,000$ level string characters for longer gaps. We can see that the number of gaps
shorter than six favors a lower number of levels and gradually increases with the number
of levels. The number of longer gaps behaves in the completely opposite way. As the number
of levels is increasing, the number of gaps in the negative reads rises faster that the number
of the gaps in the positive reads. We think
that the longer gaps are better indicator of the bad alignment as it is common for
the good alignment that there is small gap because of the noise in the underlying signal
or our discretization method shortcomings. The bigger gaps are most of the time sign of the two
unrelated level strings aligned or very uncommon unsimilarity between positive read
and its reference. In the Figure \ref{obr:res_1_nonorm} we present results for the unsmoothed
squiggles. We can see that for unsmoothed squiggles, the count of the longer gaps
is significantly bigger.

\begin{figure}
\centerline{\includegraphics[width=0.7\textwidth, height=0.4\textheight]{images/res_1}}
\caption[TODO]{We can see the number of gaps smaller than 6 and bigger than 9 per 100 and $1\,000$ characters of smoothed level
string respectively. Both of these number are stated for positive as well as negative squiggles}
\label{obr:res_1}
\end{figure}

\begin{figure}
\centerline{\includegraphics[width=0.7\textwidth, height=0.4\textheight]{images/res_1_nonorm}}
\caption[TODO]{We can see the number of gaps smaller than 6 and bigger than 9 per 100 and $1\,000$ characters of unsmoothed level
string respectively. Both of these number are stated for positive as well as negative squiggles}
\label{obr:res_1_nonorm}
\end{figure}
